File: exercism/ocaml/beer-song/beer.ml
open Core.Std

let bottles : int -> string = function
  | 0 -> "No more bottles of beer"
  | 1 -> "1 bottle of beer"
  | n -> string_of_int n ^ " bottles of beer"

let middle : int -> string = function
  | 0 -> ".\nGo to the store and buy some more, "
  | 1 -> ".\nTake it down and pass it around, "
  | n -> ".\nTake one down and pass it around, "

let verse (n: int) : string =
  let before = bottles n
  and after = bottles ((n + 99) mod 100)
  in
  String.concat [
    before;
    " on the wall, ";
    String.lowercase (before);
    middle n;
    String.lowercase (after);
    " on the wall.\n"
  ]

let rec sing ~from:(n: int) ~until:(m: int) : string =
  verse n ^ "\n" ^ (if n <= m then "" else sing ~from:(n - 1) ~until:m)

File: exercism/haskell/beer-song/Beer.hs
module Beer (sing, verse) where

import Data.Char (toUpper)

sing :: Int -> Int -> String
sing upper lower = unlines $ map verse $ reverse [lower .. upper]

verse :: Int -> String
verse n = unlines [
  upcase $ number n ++ " of beer on the wall, " ++ number n ++ " of beer.",
  middle n ++ ", " ++ number (mod (n + 99) 100) ++ " of beer on the wall."
  ]
  where
    number 0 = "no more bottles"
    number 1 = "1 bottle"
    number n = show n ++ " bottles"
    middle 0 = "Go to the store and buy some more"
    middle 1 = "Take it down and pass it around"
    middle n = "Take one down and pass it around"
    upcase (x : xs) = toUpper x : xs
